acme/domain.go

```diff
@@ -0,0 +1,247 @@
+			return "", fmt.Errorf("NS %s returned %s for %s", nameserver, dns.RcodeToString[in.Rcode], d.GetFqdn())
+		}
+		// We have a success, so one of the answers has to be a SOA RR
+		for _, ans := range in.Answer {
+			if soa, ok := ans.(*dns.SOA); ok {
+				zone := soa.Hdr.Name
+				// If we ended up on one of the TLDs, it means the domain did not exist.
+				publicsuffix, _ := publicsuffix.PublicSuffix(UnFqdn(zone))
+				if publicsuffix == UnFqdn(zone) {
+					return "", fmt.Errorf("Could not determine zone authoritatively")
+				}
+
+				d.authoritativeZone = zone
+				return d.authoritativeZone, nil
+			}
+		}
+		// Or it is NODATA, fall through to NXDOMAIN
+	}
+	// Search the authority section for our precious SOA RR
+	for _, ns := range in.Ns {
+		if soa, ok := ns.(*dns.SOA); ok {
+			zone := soa.Hdr.Name
+			// If we ended up on one of the TLDs, it means the domain did not exist.
+			publicsuffix, _ := publicsuffix.PublicSuffix(UnFqdn(zone))
+			if publicsuffix == UnFqdn(zone) {
+				return "", fmt.Errorf("Could not determine zone authoritatively")
+			}
+
+			d.authoritativeZone = zone
+			return d.authoritativeZone, nil
+		}
+	}
+	return "", fmt.Errorf("NS %s did not return the expected SOA record in the authority section", nameserver)
+}
+
+// checkDNSPropagation checks if the expected TXT record has been propagated to all authoritative nameservers.
+func (d *Domain) CheckDNSPropagation(value string) (bool, error) {
+
+	fqdn := d.GetFqdn()
+
+	// Initial attempt to resolve at the recursive NS
+	r, err := DnsQuery(fqdn, dns.TypeTXT, recursiveNameserver, true)
+	if err != nil {
+		return false, err
+	}
+	if r.Rcode == dns.RcodeSuccess {
+		// If we see a CNAME here then use the alias
+		for _, rr := range r.Answer {
+			if cn, ok := rr.(*dns.CNAME); ok {
+				if cn.Hdr.Name == fqdn {
+					fqdn = cn.Target
+					break
+				}
+			}
+		}
+	}
+
+	authoritativeNss, err := d.LookupNameservers()
+	if err != nil {
+		return false, err
+	}
+
+	return checkAuthoritativeNss(fqdn, value, authoritativeNss)
+}
+
+// LookupNameservers returns the authoritative nameservers for the given fqdn.
+func (d *Domain) LookupNameservers() ([]string, error) {
+	var authoritativeNss []string
+
+	zone, err := d.GetAuthoritativeZone()
+	if err != nil {
+		return nil, err
+	}
+
+	r, err := DnsQuery(zone, dns.TypeNS, recursiveNameserver, true)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, rr := range r.Answer {
+		if ns, ok := rr.(*dns.NS); ok {
+			authoritativeNss = append(authoritativeNss, strings.ToLower(ns.Ns))
+		}
+	}
+
+	if len(authoritativeNss) > 0 {
+		return authoritativeNss, nil
+	}
+	return nil, fmt.Errorf("Could not determine authoritative nameservers")
+}
+
+// checkAuthoritativeNss queries each of the given nameservers for the expected TXT record.
+func checkAuthoritativeNss(fqdn, value string, nameservers []string) (bool, error) {
+	for _, ns := range nameservers {
+		r, err := DnsQuery(fqdn, dns.TypeTXT, net.JoinHostPort(ns, "53"), false)
```

Comment:
I still think moving adding the port to the discovery in line 187 is a worthwhile change, as it would allow for testing of checkAuthorativeNss() with a mocked out nameserver (which will have a random port on localhost)